From 0eb42a9752a61d7767fc0b6cb3b570e8878f8fd6 Mon Sep 17 00:00:00 2001
From: Dmitry <Dmitry@espressif.com>
Date: Thu, 26 Apr 2018 11:54:14 +0300
Subject: [PATCH 2/2] Avoid assert for ecp1 register access.

---
 gdb/regcache.c    | 13 ++++++++-----
 gdb/xtensa-tdep.c | 10 +++++++++-
 2 files changed, 17 insertions(+), 6 deletions(-)

diff --git a/gdb/regcache.c b/gdb/regcache.c
index 86e648ae3d..fdecddcf36 100644
--- a/gdb/regcache.c
+++ b/gdb/regcache.c
@@ -447,11 +447,14 @@ regcache_register_status (const struct regcache *regcache, int regnum)
 {
   gdb_assert (regcache != NULL);
   gdb_assert (regnum >= 0);
-  if (regcache->readonly_p)
-    gdb_assert (regnum < regcache->descr->nr_cooked_registers);
-  else
-    gdb_assert (regnum < regcache->descr->nr_raw_registers);
-
+  /* Remove assert*/
+  /* Here some readonly registers marked as not readonly.
+     That's why value out of range.
+  */
+//  if (regcache->readonly_p)
+//    gdb_assert (regnum < regcache->descr->nr_cooked_registers);
+//  else
+//    gdb_assert (regnum < regcache->descr->nr_raw_registers);
   return regcache->register_status[regnum];
 }
 
diff --git a/gdb/xtensa-tdep.c b/gdb/xtensa-tdep.c
index 55e7d9816f..acfc0f4c84 100644
--- a/gdb/xtensa-tdep.c
+++ b/gdb/xtensa-tdep.c
@@ -2972,7 +2972,15 @@ Unable to decode Xtensa Window Interrupt Handler's code."));
   epc1_regnum = xtensa_find_register_by_name (gdbarch,"epc1");
   if (epc1_regnum < 0)
     error(_("Unable to read Xtensa register EPC1"));
-  cache->ra = xtensa_read_register (epc1_regnum);
+  
+  /* xtensa_read_register not works because try to read regsters from cache.
+    xtensa_read_register will be changed to regcache_cooked_read_unsigned to
+    read register inplace.
+  */
+  //cache->ra = xtensa_read_register (epc1_regnum);
+  regcache_cooked_read_unsigned (get_current_regcache (), regnum, &cache->ra);
+  DEBUGTRACE ("xtensa_window_interrupt_frame_cache: epc1 = 0x%08x\n", (int) cache->ra);
+  
   cache->pc = get_frame_func (this_frame);
 }
 
-- 
2.16.1.windows.4

